src/mito2/src/flush.rs

```diff
@@ -356,17 +357,39 @@ impl RegionFlushTask {
             let max_sequence = stats.max_sequence();
             series_count += stats.series_count();
 
+            let merge_mode = version.options.merge_mode.unwrap_or(MergeMode::LastRow);
+
             let source = if ranges.len() == 1 {
                 let only_range = ranges.into_values().next().unwrap();
-                Source::Iter(only_range.build_iter()?)
+                let iter = only_range.build_iter()?;
+                let iter = if merge_mode == MergeMode::LastNonNull {
+                    Box::new(LastNonNullIter::new(iter)) as _
+                } else {
+                    Box::new(iter) as _
+                };
+                Source::Iter(iter)
             } else {
                 // todo(hl): a workaround since sync version of MergeReader is wip.
                 let sources = ranges
                     .into_values()
                     .map(|r| r.build_iter().map(Source::Iter))
                     .collect::<Result<Vec<_>>>()?;
                 let merge_reader = MergeReaderBuilder::from_sources(sources).build().await?;
-                Source::Reader(Box::new(merge_reader))
+                let maybe_dedup = if version.options.append_mode {
+                    // no dedup in append mode
+                    Box::new(merge_reader) as _
+                } else {
+                    // dedup according to merge mode
+                    match version.options.merge_mode.unwrap_or(MergeMode::LastRow) {
+                        MergeMode::LastRow => {
+                            Box::new(DedupReader::new(merge_reader, LastRow::new(true))) as _
+                        }
+                        MergeMode::LastNonNull => {
+                            Box::new(DedupReader::new(merge_reader, LastNonNull::new(true))) as _
+                        }
```

Comment:
Should we filter deleted here? We may lose the deletion mark.

Reply:
fixed in https://github.com/GreptimeTeam/greptimedb/pull/6549/commits/31a6b9a6799563d67b933fcfe782ecba48b9dffb#diff-c78104950cafff2a2d1ffe864483e7b755f7b520bc9659e2798787ec590d1600L385-L389